jpyinterpreter/src/main/java/org/optaplanner/jpyinterpreter/PythonFunctionSignature.java

```diff
@@ -306,121 +333,269 @@ void unwrapBoundMethod(FunctionMetadata functionMetadata, StackMetadata stackMet
+            methodVisitor.visitJumpInsn(Opcodes.GOTO, doneGettingType);
+            methodVisitor.visitLabel(ifIsBoundFunction);
+            methodVisitor.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(BoundPythonLikeFunction.class));
+            methodVisitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, Type.getInternalName(BoundPythonLikeFunction.class),
+                    "getInstance", Type.getMethodDescriptor(Type.getType(PythonLikeObject.class)),
+                    false);
+            methodVisitor.visitLabel(doneGettingType);
+            localVariableHelper.writeTemp(methodVisitor, Type.getType(PythonLikeObject.class), argumentLocals[0]);
+        }
+
+        // Now load and typecheck the local variables
+        for (int i = 0; i < Math.min(specPositionalArgumentCount, argumentCount); i++) {
+            localVariableHelper.readTemp(methodVisitor, Type.getType(PythonLikeObject.class), argumentLocals[i]);
+            methodVisitor.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(argumentSpec.getArgumentType(i)));
+        }
+
+        // Load any arguments missing values
+        for (int i = specPositionalArgumentCount - missingValues; i < specPositionalArgumentCount; i++) {
+            if (argumentSpec.isArgumentNullable(i)) {
+                methodVisitor.visitInsn(Opcodes.ACONST_NULL);
+            } else {
+                methodVisitor.visitFieldInsn(Opcodes.GETSTATIC, Type.getInternalName(defaultArgumentHolderClass),
+                        PythonDefaultArgumentImplementor.getConstantName(i),
+                        Type.getDescriptor(argumentSpec.getArgumentType(i)));
+            }
         }
 
-        if (argumentCount > descriptorParameterTypes.length) {
+        // Load *vargs and **kwargs if the function has them
+        if (argumentSpec.hasExtraPositionalArgumentsCapture()) {
             localVariableHelper.readTemp(methodVisitor, Type.getType(PythonLikeTuple.class),
-                    argumentLocals[extraPositionalArgumentsVariableIndex.get()]);
-            for (int i = argumentCount; i >= descriptorParameterTypes.length; i--) {
-                methodVisitor.visitInsn(Opcodes.DUP_X1);
+                    capturedExtraPositionalArgumentsLocal);
+        }
+
+        if (argumentSpec.hasExtraKeywordArgumentsCapture()) {
+            // No kwargs for call method, so just load an empty map
+            CollectionImplementor.buildMap(PythonLikeDict.class, methodVisitor, 0);
+        }
+
+        // Call the method
+        methodDescriptor.callMethod(methodVisitor);
+
+        // Free temporary locals for arguments
+        for (int i = 0; i < argumentLocals.length; i++) {
+            localVariableHelper.freeLocal();
+        }
+        // Free temporary local for vargs
+        localVariableHelper.freeLocal();
+    }
+
+    public void callPython311andAbove(FunctionMetadata functionMetadata, StackMetadata stackMetadata, int argumentCount,
+            List<String> keywordArgumentNameList) {
```

Comment:
Looking at this from the point of view of separation of concerns and reducing the complexity of the class, why not create a specialized variant of this class that handles Python 3.11? (3.10, 3.9, ...)

That way, version-specific complexity is cleanly isolated and the commonalities are more obvious, as they are in a parent class.